compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.test.cpp

```diff
@@ -0,0 +1,395 @@
+   * (1, 192, 196)
+   *      C   AB
+   *      |
+   * (1, 196, 192)
+   *      AB   C
+   */
+  g.init(/*in*/ {1, 14, 14, 192}, /*perm*/ {0, 3, 1, 2}, /*reshape*/ {1, 192, 196},
+         /*perm*/ {0, 2, 1}, /*out*/ {1, 196, 192});
+
+  EXPECT_TRUE(pass.run(g.g()));
+}
+
+TEST(RemoveUnnecessaryTransposeNetPass, rank_reduction_pattern2)
+{
+  TransposeReshapeTransposeGraph g;
+  luci::RemoveUnnecessaryTransposeNetPass pass;
+
+  /** from detr
+   *
+   * (1, 7, 7, 448)
+   *     A  B   C
+   *      |
+   * (1, 448, 7, 7)
+   *      C   A  B
+   *       |
+   * (1, 448, 49)
+   *      C   AB
+   *      |
+   * (1, 49, 448)
+   *     AB   C
+   */
+  g.init(/*in*/ {1, 7, 7, 448}, /*perm*/ {0, 3, 1, 2}, /*reshape*/ {1, 448, 49}, /*perm*/ {0, 2, 1},
+         /*out*/ {1, 49, 448});
+
+  EXPECT_TRUE(pass.run(g.g()));
+}
+
+TEST(RemoveUnnecessaryTransposeNetPass, rank_expand_pattern1)
+{
+  TransposeReshapeTransposeGraph g;
+  luci::RemoveUnnecessaryTransposeNetPass pass;
+
+  /**
+   * (10, 15, 20)
+   *  A   B   C
+   *      |           => transpose
+   * (10, 20, 15)
+   *  A   C   B
+   *      |           => reshape
+   * (10, 20, 3, 5)
+   *  A   C   Ba Bb
+   *      |           => transpose
+   * (10, 3, 5, 20)
+   *  A   Ba Bb C
+   */
+
+  g.init(/*in*/ {10, 15, 20}, /*perm*/ {0, 2, 1}, /*reshape*/ {10, 20, 3, 5}, /*perm*/ {0, 2, 3, 1},
+         /*out*/ {10, 3, 5, 20});
+
+  EXPECT_TRUE(pass.run(g.g()));
+}
+
+TEST(RemoveUnnecessaryTransposeNetPass, rank_expand_with_one_pattern)
+{
+  TransposeReshapeTransposeGraph g;
+  luci::RemoveUnnecessaryTransposeNetPass pass;
+
+  /**
+   * (20, 50, 100)
+   *      |           => transpose
+   * (20, 100, 50)
+   *      |           => reshape
+   * (2, 1, 10, 10, 10, 5, 10, 1)
+   *      |           => transpose
+   * (1, 1, 2, 10, 5, 10, 10, 10)
+   */
+
+  g.init(/*in*/ {20, 50, 100}, /*perm*/ {0, 2, 1}, /*reshape*/ {2, 1, 10, 10, 10, 5, 10, 1},
+         /*perm*/ {7, 1, 0, 2, 5, 6, 3, 4},
+         /*out*/ {1, 1, 2, 10, 5, 10, 10, 10});
+
+  EXPECT_TRUE(pass.run(g.g()));
+}
+
+TEST(RemoveUnnecessaryTransposeNetPass, rank_reduction_with_one_pattern)
+{
+  TransposeReshapeTransposeGraph g;
+  luci::RemoveUnnecessaryTransposeNetPass pass;
+
+  /**
+   * (1, 100, 10, 12)
+   *      A   B   C
+   *        |
+   * (1, 10, 12, 100)
+   *     B   C   A
+   *        |
+   *    (120, 100)
+   *      BC   A
+   *        |
+   *    (100, 120)
+   *      A   BC
+   */
+  g.init(/*in*/ {1, 100, 10, 12}, /*perm*/ {0, 2, 3, 1}, /*reshape*/ {120, 100}, /*perm*/ {1, 0},
+         /*out*/ {100, 120});
+
+  EXPECT_TRUE(pass.run(g.g()));
+}
+
+TEST(RemoveUnnecessaryTransposeNetPass, complex_reshape_pattern1)
+{
+  TransposeReshapeTransposeGraph g;
+  luci::RemoveUnnecessaryTransposeNetPass pass;
+
+  /**
+   * (12, 12, 15, 2)
+   *   A  B   C   D
+   *      |
+   * (2, 12, 12, 15)
+   *  D  A   B   C
+   *      |
+   * (2, 36, 12, 5)
+   *  D  ABCa ABCb ABCc
+   *      |
+   * (36, 12, 5, 2)
+   * ABCa ABCb ABCc D
```

Comment:
How `tags` working, I think this example might be helpful.

Reply:
Q) is there any particular reason using ABCD ?
for Tensors, we usually use HW, CHW nad NCHW and I think you already know this.